pkg/registry/apps/statefulset/storage/storage_test.go

```diff
@@ -255,43 +292,80 @@ func TestScaleUpdate(t *testing.T) {
-		},
-		Spec: extensions.ScaleSpec{
-			Replicas: int32(replicas),
-		},
-	}
-
-	if _, _, err := storage.Scale.Update(ctx, update.Name, rest.DefaultUpdatedObjectInfo(&update), rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc); err != nil {
-		t.Fatalf("error updating scale %v: %v", update, err)
-	}
-
-	obj, err := storage.Scale.Get(ctx, name, &metav1.GetOptions{})
-	if err != nil {
-		t.Fatalf("error fetching scale for %s: %v", name, err)
-	}
-	scale := obj.(*extensions.Scale)
-	if scale.Spec.Replicas != int32(replicas) {
-		t.Errorf("wrong replicas count expected: %d got: %d", replicas, scale.Spec.Replicas)
-	}
-
-	update.ResourceVersion = sts.ResourceVersion
-	update.Spec.Replicas = 15
-
-	if _, _, err = storage.Scale.Update(ctx, update.Name, rest.DefaultUpdatedObjectInfo(&update), rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc); err != nil && !errors.IsConflict(err) {
-		t.Fatalf("unexpected error, expecting an update conflict but got %v", err)
+	versions := []string{"v1beta1", "v1beta2", "v1"}
+	for _, version := range versions {
+		name := "foo-" + version
+		var sts apps.StatefulSet
+		ctx := genericapirequest.WithNamespace(genericapirequest.NewContext(), metav1.NamespaceDefault)
+		requestInfo := genericapirequest.RequestInfo{
+			APIGroup:   "apps",
+			APIVersion: version,
+			Resource:   "statefulsets",
+		}
+		ctx = genericapirequest.WithRequestInfo(ctx, &requestInfo)
+		key := "/statefulsets/" + metav1.NamespaceDefault + "/" + name
+		validStatefulSet.Name = name
+		if err := storage.StatefulSet.Storage.Create(ctx, key, &validStatefulSet, &sts, 0); err != nil {
+			t.Fatalf("error setting new statefulset (key: %s) %v: %v", key, validStatefulSet, err)
+		}
+		replicas := 12
+		extensionsScale := extensions.Scale{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: metav1.NamespaceDefault,
+			},
+			Spec: extensions.ScaleSpec{
+				Replicas: int32(replicas),
+			},
+		}
+		autoScale := autoscaling.Scale{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: metav1.NamespaceDefault,
+			},
+			Spec: autoscaling.ScaleSpec{
+				Replicas: int32(replicas),
+			},
+		}
+
+		switch version {
+		case "v1beta1", "v1beta2":
+			if _, _, err := storage.Scale.Update(ctx, extensionsScale.Name, rest.DefaultUpdatedObjectInfo(&extensionsScale), rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc); err != nil {
+				t.Fatalf("error updating scale %v: %v", extensionsScale, err)
+			}
+		default:
+			if _, _, err := storage.Scale.Update(ctx, autoScale.Name, rest.DefaultUpdatedObjectInfo(&autoScale), rest.ValidateAllObjectFunc, rest.ValidateAllObjectUpdateFunc); err != nil {
+				t.Fatalf("error updating scale %v: %v", autoScale, err)
+			}
+		}
+
+		obj, err := storage.Scale.Get(ctx, name, &metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("error fetching scale for %s: %v", name, err)
+		}
+
+		switch version {
+		case "v1beta1", "v1beta2":
```

Comment:
There is a lot of redundant codes due to handling different group versions separately. An idea is to create a generic helper function. I wish to get first round of review before working on the generic helper function. This is to ensure I am on the right track and facilitate code review.